### PR TITLE
Don't reuse registers in Debug mode

### PR DESCRIPTION
**Pull request by EgorBo**
Example:
```csharp
using System;

public class Program
{
    static void Main()
    {
        int a = 0;
        sumcheck(0);
        Console.ReadLine();
    }

    static void sumcheck(int val) => Console.WriteLine(val);
}
```
IL code for `Main`:
```
        IL_0000: nop
        IL_0001: ldc.i4.0
        IL_0002: stloc.0   <-- we save 0 to a loc0, but sumcheck is called with its own ldc.i4.0:
        IL_0003: ldc.i4.0  <--- 
        IL_0004: call void Program::sumcheck(int32)
        IL_0009: nop
        IL_000a: call string [System.Console]System.Console::ReadLine()
        IL_000f: pop
        IL_0010: ret
```
Codegen diff for `Main` under EnC with my change:
```diff
G_M27646_IG01:
       55                   push     rbp
       4883EC30             sub      rsp, 48
       488D6C2430           lea      rbp, [rsp+30H]
       33C0                 xor      eax, eax
       8945FC               mov      dword ptr [rbp-04H], eax
       488945F0             mov      qword ptr [rbp-10H], rax
G_M27646_IG02:              ;; offset=0013H
       833DAEFD240000       cmp      dword ptr [(reloc 0x7ffc4adcc858)], 0
       7405                 je       SHORT G_M27646_IG04
G_M27646_IG03:              ;; offset=001CH
       E83F9C685F           call     CORINFO_HELP_DBG_IS_JUST_MY_CODE
G_M27646_IG04:
       90                   nop
       33C9                 xor      ecx, ecx
       894DFC               mov      dword ptr [rbp-04H], ecx
+      33C9                 xor      ecx, ecx
       E82A5CFFFF           call     Program:sumcheck(int)
       90                   nop
       E85CFCFFFF           call     System.Console:ReadLine():System.String
       488945F0             mov      gword ptr [rbp-10H], rax
       90                   nop
       90                   nop
G_M27646_IG05:
       4883C430             add      rsp, 48
       5D                   pop      rbp
       C3                   ret
```
Potentially, it may lead to a case where after we drag the "debug arrow" back to sumcheck() after Console.ReadLine() we might end up with a garbage in ecx
image

@dotnet/jit-contrib @kunalspathak PTAL - do I disable SetReuseRegVal correctly?

### CODE REVIEW

**Comment by kunalspathak** [comment]
> after we drag the "debug arrow" back to sumcheck() after Console.ReadLine() we might end up with a garbage in ecx

Do we know if the impact will be visible to the user? In your example, I don't think that would be the case.

**Comment by EgorBo** [comment]
> > after we drag the "debug arrow" back to sumcheck() after Console.ReadLine() we might end up with a garbage in ecx
> 
> Do we know if the impact will be visible to the user? In your example, I don't think that would be the case.

This problem (screenshot) reproduces on .NET Framework x64 (32bit is fine) and I guess it can reproduce on CoreCLR too, just not for this specific repro since Console.WriteLine would clear ecx itself

**Comment by EgorBo** [comment]
@kunalspathak some runtime tests are marked as 
```
<DebugType>Full</DebugType>
<Optimize>False</Optimize>
```
so when we replay MCH we compile methods in them without optimizations and with that debug flag so it's expected.
Example: Test_HndIndex_10_Plain

File: src/coreclr/jit/lsra.cpp [hunk 1]
```diff
@@ -2808,7 +2808,7 @@ regNumber LinearScan::allocateReg(Interval*    currentInterval,
             bool wasAssigned = regSelector->foundUnassignedReg() && (assignedInterval != nullptr) &&
                                (assignedInterval->physReg == foundReg);
             unassignPhysReg(availablePhysRegRecord ARM_ARG(currentInterval->registerType));
-            if (regSelector->isMatchingConstant())
+            if (regSelector->isMatchingConstant() && !compiler->opts.compDbgCode)
```
Comment (BruceForstall) [inline src/coreclr/jit/lsra.cpp:2811]:
```suggestion
            if (regSelector->isMatchingConstant() && compiler->OptimizationEnabled())
```
?

**Comment by kunalspathak** [comment]
> so when we replay MCH we compile methods in them without optimizations and with that debug flag so it's expected.
Example: Test_HndIndex_10_Plain

Not sure if I follow. Because of presence of those flags in `.csproj`, we compile the `dll` in Debug-mode, optimizations disabled and do the collection. When we replay, we would get same method contexts of debug-mode methods, right? I know I am missing something here.

**Comment by EgorBo** [comment]
> > so when we replay MCH we compile methods in them without optimizations and with that debug flag so it's expected.
> > Example: Test_HndIndex_10_Plain
> 
> Not sure if I follow. Because of presence of those flags in `.csproj`, we compile the `dll` in Debug-mode, optimizations disabled and do the collection. When we replay, we would get same method contexts of debug-mode methods, right? I know I am missing something here.

that is my understanding as well, coreclr_tests suite is full of those compiled/replayed without optimizations, a good example is @SingleAccretion's PR https://github.com/dotnet/runtime/pull/63957 with a lot of regressions in that collection

**Comment by kunalspathak** [comment]
> that is my understanding as well, coreclr_tests suite is full of those compiled/replayed without optimizations

Are you saying that is the reason we are seeing missing value? I was asking if you analyzed the regressions that are coming out of this change specially in `coreclr_tests` ? These are the diffs from the previous commit. Might have to wait and see if the change of `OptimizationsEnabled()` has any impact.


```
coreclr_tests.pmi.Linux.arm64.checked.mch:

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 168370616 (overridden on cmd)
Total bytes of diff: 168457140 (overridden on cmd)
Total bytes of delta: 86524 (0.05 % of base)
    diff is a regression.
    relative diff is a regression.

coreclr_tests.pmi.Linux.x64.checked.mch:

Summary of Code Size diffs:
(Lower is better)

Total bytes of base: 126740366 (overridden on cmd)
Total bytes of diff: 126760651 (overridden on cmd)
Total bytes of delta: 20285 (0.02 % of base)
    diff is a regression.
    relative diff is a regression.
```

**Review by kunalspathak — APPROVED** [review]
LGTM

**Pull request merged**